### PR TITLE
SAK-44622: No bullhorn notification is thrown when an announcement's …

### PR DESCRIPTION
**Pull request by JKJANETSCHEK**
…Beginning Date is in the future

### CODE REVIEW

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java [hunk 1]
```diff
@@ -123,7 +123,10 @@ public interface AnnouncementService extends MessageService
     public static final String EVENT_ANNC_UPDATE_AVAILABILITY = SECURE_ANNC_ROOT + "revise.availability";
 
     public static final String MOTD_TOOL_ID = "sakai.motd";
-    
+
+	/** SAK-44622 **/
+	public static final String EVENT_AVAILABLE_ANNC = SECURE_ANNC_ROOT + "available.announcement";
```
Comment (adrianfish) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:128]: Indentation is off here.

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsBullhornHandler.java [hunk 2]
```diff
@@ -70,7 +70,7 @@ public class AnnouncementsBullhornHandler extends AbstractBullhornHandler {
 
     @Override
     public List<String> getHandledEvents() {
-        return Arrays.asList(AnnouncementService.SECURE_ANNC_ADD, AnnouncementService.EVENT_ANNC_UPDATE_AVAILABILITY, AnnouncementService.SECURE_ANNC_REMOVE_OWN, AnnouncementService.SECURE_ANNC_REMOVE_ANY);
+        return Arrays.asList(AnnouncementService.SECURE_ANNC_ADD, AnnouncementService.EVENT_ANNC_UPDATE_AVAILABILITY, AnnouncementService.SECURE_ANNC_REMOVE_OWN, AnnouncementService.SECURE_ANNC_REMOVE_ANY,AnnouncementService.EVENT_AVAILABLE_ANNC);
```
Comment (adrianfish) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementsBullhornHandler.java:73]:
```suggestion
        return Arrays.asList(AnnouncementService.SECURE_ANNC_ADD, AnnouncementService.EVENT_ANNC_UPDATE_AVAILABILITY, AnnouncementService.SECURE_ANNC_REMOVE_OWN, AnnouncementService.SECURE_ANNC_REMOVE_ANY, AnnouncementService.EVENT_AVAILABLE_ANNC);
```

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 3]
```diff
@@ -2863,8 +2864,30 @@ else if (announceTo != null && announceTo.equals("groups"))
 					{
 						// availablity changed
 						eventTrackingService.post(eventTrackingService.newEvent(AnnouncementService.EVENT_ANNC_UPDATE_AVAILABILITY, msg.getReference(), true));
+
+						//SAK-44622:
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2868]: I wouldn't bother with the SAK number, to be honest. We can backtrack using git blame anyway, if we need to look up which JIRA introduced a change.

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java [hunk 4]
```diff
@@ -123,7 +123,10 @@ public interface AnnouncementService extends MessageService
     public static final String EVENT_ANNC_UPDATE_AVAILABILITY = SECURE_ANNC_ROOT + "revise.availability";
 
     public static final String MOTD_TOOL_ID = "sakai.motd";
-    
+
+	/** SAK-44622 **/
```
Comment (adrianfish) [inline announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:127]: We don't need the SAK number. We can use the git history to look that up if we need to.

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 5]
```diff
@@ -2863,8 +2864,30 @@ else if (announceTo != null && announceTo.equals("groups"))
...
+						//check if an delay might exist
+						if ((StringUtils.isNotEmpty(oReleaseDate) && releaseDate == null) || (StringUtils.isNotEmpty(oReleaseDate) && releaseDate != null && !oReleaseDate.equals(releaseDate.toString()))) {
+
+							eventTrackingService.cancelDelays(msg.getReference(), org.sakaiproject.announcement.api.AnnouncementService.EVENT_AVAILABLE_ANNC);
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2872]: You don't need the fully qualified package here, just AnnouncementService will do.

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 6]
```diff
@@ -2863,8 +2864,30 @@ else if (announceTo != null && announceTo.equals("groups"))
...
+
+							//check if new date is in future
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2874]: Shouldn't this be "check if new date has passed already" ? Or something like that?

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 7]
```diff
@@ -2863,8 +2864,30 @@ else if (announceTo != null && announceTo.equals("groups"))
...
+								eventTrackingService.delay(event,msg.getHeader().getInstant());
+							}
+						}
 					}
 				}
+				//SAK-44622:
+				//Create delay
+				Instant nowInstant = Instant.now();
+				Instant date = msg.getHeader().getInstant();
+				if (nowInstant.isBefore(date)){
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2886]:
```suggestion
				Instant date = msg.getHeader().getInstant();
				if (date.isAfter(Instant.now()) {
```

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 8]
```diff
@@ -2967,6 +2990,13 @@ public void doDelete(RunData rundata, Context context)
 					//AnnouncementMessageEdit edit = channel.editAnnouncementMessage(message.getId());
 					//channel.removeMessage(edit); 
 					channel.removeAnnouncementMessage(message.getId());
+
+					//SAK-44622
+					//Delete possible delay
+					Instant date = message.getHeader().getInstant();
+					if(date.isAfter(Instant.now())){
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2997]:
```suggestion
					if (message.getHeader().getInstant().isAfter(Instant.now())) {
```

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 9]
```diff
@@ -2967,6 +2990,13 @@ public void doDelete(RunData rundata, Context context)
 					//AnnouncementMessageEdit edit = channel.editAnnouncementMessage(message.getId());
 					//channel.removeMessage(edit); 
 					channel.removeAnnouncementMessage(message.getId());
+
+					//SAK-44622
```
Comment (adrianfish) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2994]: I'd remove the SAK number, we can get that via git history.

**Review by adrianfish — CHANGES_REQUESTED** [review]
Some small changes. Thanks for the contribution.

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java (hunk 4 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:127]: done

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java (hunk 1 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:128]: I am not exactly sure what you mean

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 3 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2868]: done

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 5 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2872]: done

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 6 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2874]: done

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 7 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2886]: done

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 9 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2994]: done

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 8 above)
Comment (JKJANETSCHEK) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:2997]: done

File: announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java (hunk 1 above)
Comment (adrianfish) [reply announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:128]: the formatting seems a bit weird, not lined up with line 125

Comment (JKJANETSCHEK) [reply announcement/announcement-api/api/src/java/org/sakaiproject/announcement/api/AnnouncementService.java:128]: done, realigned line

**Review by adrianfish — APPROVED** [review]

**Review by ern — APPROVED** [review]

**Pull request merged**